// Copyright (C) 2018 Bluzelle
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

syntax = "proto3";

import "database.proto";

message bzn_envelope
{
    string sender = 1;
    bytes signature = 2;
    uint64 timestamp = 3;

    oneof payload
    {
        bytes database_msg = 10;
        bytes pbft_internal_request = 11;

        bytes database_response = 12;
        bytes json = 13;
        bytes audit = 14;
        bytes pbft = 15;
        bytes pbft_membership = 16;
        bytes status_request = 17;
    }
}

message bzn_msg
{
    // Keeping this around for raft
    oneof msg
    {
        database_msg db = 10;
        string json = 11;
    }
}

enum bzn_msg_type
{
    BZN_MSG_UNDEFINED = 0;
    BZN_MSG_PBFT = 1;
    BZN_MSG_PBFT_MEMBERSHIP = 2;
}

message wrapped_bzn_msg
{
    // This is stored as a serialized string because we need to sign it, and serialization is not guarenteed to be deterministic
    string payload = 1;
    bzn_msg_type type = 2;

    string sender = 3;
    bytes signature = 4;
}
